### DISCUSSION TITLE
Improved media support

### DISCUSSION THREAD

**Discussion by Bertg** [remirror/remirror#458] [Ideas]
When adding images, we there is an expectation that the image displays nicely inline. However, this might not always be the case. For example, someone might have a series of screenshots they want to attach, without these being displayed inline. 

On the flip side, a user may attach a single paged PDF, that they would like to have displayed inline. 

At it's core, we are dealing with files that need to be uploaded to a backend, and - possibly - be displayed inline. 

Here are some loose ideas on features that could be interesting to discuss:

- Create a module that will upload a file. This module itself can be configured with components that can preview multiple items. A previewer for images, PDF's, Audio, .... This module is also responsible to provide a preview during upload. The file module can contain overarching features like annotations.
- Display file placeholder. In case no preview can be generated (or the user chooses not to have a preview) a placeholder is displayed instead. The annotation value can be included in this design.
- Display multiple "placeholders" as a list or grid. When multiple files are stored in sequence we can optimise the rendering to show them as one list, or even provide an option to the user to show these items in a grid.
- When displaying as a placeholder, the preview renderer can still provide a nice icon.

If there is interest, I can work on the UX for this, as I have some experience on how to do this flow.

**Comment by ifiokjr** [comment]
@Bertg this sounds like a really useful feature.

Let's define everything that can attach a file as `media`. i.e. the following are all media.

- images (with and without captions)
- video
- pdfs
- powerpoint slides
- audio
- basically any file type

Most of these will have their own separate extensions but they are all tied together under that `media` banner.

All media will have the following attribute available via the `node.attrs.displayMode`.

```ts
type MediaDisplayMode = 'inline' | 'placeholder' | string;
```

As you mention, at the moment every image assumes that it should be displayed `inline`. For flexibility, developers using remirror need to be able to configure the default display mode when a media item is dropped into the editor. This display mode should be able to respond to the way the file was added. 

If you click the add image button then the `displayMode` should probably be `inline`. If drop a single image, maybe still inline. If you drop 10 images then probably a placeholder for each image.

This raises an interesting problem. What if the user drops multiple files of different types. A bundle of pdf's, images, or videos. How should these be handled by the editor?

> If there is interest, I can work on the UX for this, as I have some experience on how to do this flow.

There's definitely interest from me 😄  and if you're happy to work on the UI, I'll make the codebase ready for the integration.

**Comment by Bertg** [comment]
Some more loos thoughts - i don't have a lot of time to make them more concise at the moment.

General: 

* Each type of file should be able to have a caption, it should not be reserved for images.
* Captions should be optional and support basic formatting (B/I/U, links...)
* The preview itself should be convertible into a link (maybe this should be an image only feature)

Display modes:

* Inline: If there is an ability to render the file, based on the file mime-type, we render that preview. This is "full width" by default.
* Placeholder: Display an icon, filename and caption. This is "full width" by default. Several placeholders in succession should display as a single list.
![Screenshot 2020-08-14 at 10 01 45](https://user-images.githubusercontent.com/1434/90227653-37332680-de15-11ea-858d-0b485320c314.png)
* Link: The file is just a link inside the text. The link text is modifiable. Captions are not supported in this mode. (If this is configured in an editor, the "Edit link" UI should also provide an option to attach a file.

Drag & Drop default behaviour:

* If a single item is dragged in, and we can render a preview, it will be added as a inline. It is possible that during upload nothing can be rendered (in case of video, or PDF) in that case the preview must render a skeleton.
* If multiple items are dragged in and all of these are images, it will be added as a inline. 
* In al other cases items are added as placeholder

Upload via interface:

* The button or action chosen by the user, will decide on the preview method. 

Upload feedback:

* Each display mode will provide some feedback during the upload. 
* Inline: Overlay the preview with upload progress. If the upload fails, this overlay will show that information.
* Placeholder: Upload progress can be shown via a circular progress over the icon. Or a small progress bar at the top/bottom of the placeholder, or as a background colour filling from left to right. If the upload fails, the preview will display this information.
* Link: A small inline circular progress indictor is show together with the word "uploading". In case the upload fails, the progress indicator is replaced with "upload failed" if this is clicked, we open the "edit link" UI.

Editing:

* it should be possible to switch between each display mode

**Reply by ifiokjr** [reply]
I really, really like this. I've been focusing on bug fixes for the past few days, so I haven't had much time to respond. Maybe we can start working on it from midway through next week?

> The preview itself should be convertible into a link (maybe this should be an image only feature)

I wasn't quite sure what you meant by this. Do you mean that it would need to extract the URL and display the link to users?

**Reply by Bertg** [reply]
I meant that the image itself should be clickable as a link, when not in edit mode.

**Reply by ifiokjr** [reply]
Oh sure, that's possible.

**Comment by ifiokjr** [comment]
By the way, I've created a [milestone](https://github.com/remirror/remirror/milestone/4) to keep track of progress on this and related media issues.

**Comment by ronnyroeller** [comment]
Those are great requirements! To add, here two other ones we stumbled over:

## Gallery block

Would you envision to have one file per block or a "gallery block" with multiple files?

Some inspiration below:

### One file per block (ex. Atlassian)

Confluence shows one image/file per node. I also really like their resizing. Allowing to resize only horizontally feels quite smooth and avoid the frustration for users (who in 99% of cases want to keep aspect ratio):

![image](https://user-images.githubusercontent.com/9339055/91545629-928b0b80-e921-11ea-87a9-3f1c338a4256.png)

### Gallery block (ex. Medium)

Medium automatically creates a gallery block if you add multiple photos. There is no control about the sizing of photos.

#### One photo
![image](https://user-images.githubusercontent.com/9339055/91546275-7c317f80-e922-11ea-962b-4b8a368736a8.png)

#### Two photos
![image](https://user-images.githubusercontent.com/9339055/91546236-6cb23680-e922-11ea-9e2a-c9bf2914ef38.png)

#### Three photos
![image](https://user-images.githubusercontent.com/9339055/91546126-47bdc380-e922-11ea-93e2-560c046b9ec1.png)

## Wrapping text

Users should have the option to let text flow around the file.

Here an example from Atlassian Confluence:
![image](https://user-images.githubusercontent.com/9339055/91546617-0843a700-e923-11ea-92ac-621706b4b65c.png)

**Comment by Bertg** [comment]
I like the idea of galleries. But I wouldn't make the automatically. 

The simplest of which I already described above: 

> Display multiple "placeholders" as a list or grid. When multiple files are stored in sequence we can optimise the rendering to show them as one list, or even provide an option to the user to show these items in a grid.

I think it could be cool to also have multiple images either show as sequence of blocks (default) or as a gallery, mosaic, slideshow. I'm not sure how easy or difficult this would be. I think it should be possible to add data to each node indicating if they should form a gallery (or other) with the nodes around them. However how that would work interface wise I have no idea.
